mlflow/types/type_hints.py

```diff
@@ -0,0 +1,309 @@
+        # but added here to avoid potential edge cases
+        if annotation is None:
+            invalid_fields.append(field_name)
+            continue
+        colspec_type = _infer_colspec_type_from_type_hint(annotation)
+        if colspec_type.required is False and field_required(field_info):
+            raise MlflowException.invalid_parameter_value(
+                f"Optional field `{field_name}` in Pydantic model `{model.__name__}` "
+                "doesn't have a default value. Please set default value to None for this field."
+            )
+        if field_type == "Property":
+            output_fields.append(
+                Property(
+                    name=field_name,
+                    dtype=colspec_type.dtype,
+                    required=colspec_type.required,
+                )
+            )
+        elif field_type == "ColSpec":
+            output_fields.append(
+                ColSpec(
+                    type=colspec_type.dtype,
+                    name=field_name,
+                    required=colspec_type.required,
+                )
+            )
+    if invalid_fields:
+        raise MlflowException.invalid_parameter_value(
+            "The following fields in the Pydantic model do not have type annotations: "
+            f"{invalid_fields}. Please add type annotations to these fields."
+        )
+
+    return output_fields
+
+
+def _is_pydantic_type_hint(type_hint: type[Any]) -> bool:
+    try:
+        return inspect.isclass(type_hint) and issubclass(type_hint, pydantic.BaseModel)
+    # inspect.isclass(dict[str, int]) is True, but issubclass raises a TypeError
+    except TypeError:
+        return False
+
+
+def model_fields(
+    model: pydantic.BaseModel,
+) -> dict[str, type[FIELD_TYPE]]:
+    if PYDANTIC_V1_OR_OLDER:
+        return model.__fields__
+    return model.model_fields
+
+
+def model_validate(model: pydantic.BaseModel, values: Any) -> None:
+    if PYDANTIC_V1_OR_OLDER:
+        model.validate(values)
+    else:
+        # use strict mode to avoid any data conversion here
+        # e.g. "123" will not be converted to 123 if the type is int
+        model.model_validate(values, strict=True)
+
+
+def field_required(field: type[FIELD_TYPE]) -> bool:
+    if PYDANTIC_V1_OR_OLDER:
+        return field.required
+    return field.is_required()
+
+
+def _infer_schema_from_type_hint(type_hint: type[Any]) -> Schema:
+    if _is_pydantic_type_hint(type_hint):
+        col_specs = _infer_fields_from_pydantic_model(type_hint, "ColSpec")
+        return Schema(inputs=col_specs)
+    else:
+        col_spec_type = _infer_colspec_type_from_type_hint(type_hint)
+        # Creating Schema with unnamed optional inputs is not supported
+        if col_spec_type.required is False:
+            raise MlflowException.invalid_parameter_value(
+                "Optional type hint is not supported, it can only be used as a field type in "
+                "Pydantic models."
```

Comment:
Can this be updated to provide corrective guidance to users? "If you would like to provide Optional types, use a Pydantic-based type hint definition <link to docs section?>"

Reply:
I'll add the doc link after type hints is added in pyfunc models